src/utilities/buildASTSchema.js

```diff
@@ -164,57 +165,31 @@ export function buildASTSchema(ast: DocumentNode): GraphQLSchema {
     }
   }
 
-  let queryTypeName;
-  let mutationTypeName;
-  let subscriptionTypeName;
+  let operationTypes = {
+    query: nodeMap.Query && 'Query',
+    mutation: nodeMap.Mutation && 'Mutation',
+    subscription: nodeMap.Subscription && 'Subscription',
+  };
+
   if (schemaDef) {
+    operationTypes = {};
     schemaDef.operationTypes.forEach(operationType => {
       const typeName = operationType.type.name.value;
-      if (operationType.operation === 'query') {
-        if (queryTypeName) {
-          throw new Error('Must provide only one query type in schema.');
-        }
-        if (!nodeMap[typeName]) {
-          throw new Error(
-            `Specified query type "${typeName}" not found in document.`
-          );
-        }
-        queryTypeName = typeName;
-      } else if (operationType.operation === 'mutation') {
-        if (mutationTypeName) {
-          throw new Error('Must provide only one mutation type in schema.');
-        }
-        if (!nodeMap[typeName]) {
-          throw new Error(
-            `Specified mutation type "${typeName}" not found in document.`
-          );
-        }
-        mutationTypeName = typeName;
-      } else if (operationType.operation === 'subscription') {
-        if (subscriptionTypeName) {
-          throw new Error('Must provide only one subscription type in schema.');
-        }
-        if (!nodeMap[typeName]) {
-          throw new Error(
-            `Specified subscription type "${typeName}" not found in document.`
-          );
-        }
-        subscriptionTypeName = typeName;
+      const operation = operationType.operation;
+
+      if (operationTypes[operation]) {
+        throw new Error(`Must provide only one ${operation} type in schema.`);
+      }
+      if (!nodeMap[typeName]) {
+        throw new Error(
+          `Specified ${operation} type "${typeName}" not found in document.`
+        );
       }
+      operationTypes[operation] = typeName;
     });
-  } else {
-    if (nodeMap.Query) {
-      queryTypeName = 'Query';
-    }
-    if (nodeMap.Mutation) {
-      mutationTypeName = 'Mutation';
-    }
-    if (nodeMap.Subscription) {
-      subscriptionTypeName = 'Subscription';
-    }
   }
 
-  if (!queryTypeName) {
+  if (!operationTypes['query']) {
```

Comment:
`Query`?
Also, we should probably create constants to avoid this type of typos.